src/Storages/StorageURL.cpp

```diff
@@ -858,6 +863,54 @@ FormatSettings StorageURL::getFormatSettingsFromArgs(const StorageFactory::Argum
     return format_settings;
 }
 
+ASTs::iterator StorageURL::collectHeaders(
+    ASTs & url_function_args, URLBasedDataSourceConfiguration & configuration, ContextPtr context)
+{
+    ASTs::iterator headers_it;
+
+    for (auto arg_it = url_function_args.begin(); arg_it != url_function_args.end(); ++arg_it)
+    {
+        const auto * headers_ast_function = (*arg_it)->as<ASTFunction>();
+        if (headers_ast_function && headers_ast_function->name == "headers")
+        {
+            if (headers_it != ASTs::iterator())
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "URL table function can have only one key-value argument: headers=(). {}",
+                    bad_arguments_error_message);
+
+            const auto * headers_function_args_expr = assert_cast<const ASTExpressionList *>(headers_ast_function->arguments.get());
+            auto headers_function_args = headers_function_args_expr->children;
+
+            for (auto & header_arg : headers_function_args)
+            {
+                const auto * header_ast = header_arg->as<ASTFunction>();
+                if (!header_ast)
```

Comment:
I guess we also should check that its function `=`, because by now we can write smth like
` select * from url(..., headers(arrayJoin('X-ClickHouse-Format', 'JSONEachRow')));`
and it will work